common/workunit/workunit.cpp

```diff
@@ -2988,10 +2971,20 @@ class CDaliWorkUnitFactory : public CWorkUnitFactory, implements IDaliClientShut
                 case WUStateDebugRunning:
                 case WUStateBlocked:
                 case WUStateAborting:
-                    SessionId agent = conn->queryRoot()->getPropInt64("@agentSession", -1);
-                    if (checkAbnormalTermination(wuid, ret, agent))
+                    if (queryDaliServerVersion().compare("2.1")>=0)
                     {
-                        return ret;
+                        SessionId agent = conn->queryRoot()->getPropInt64("@agentSession", -1);
+                        if((agent>0) && querySessionManager().sessionStopped(agent, 0))
+                        {
+                            conn->reload();
+                            // MCK - does checking state after sessionStopped reveal meaningful info ?
```

Comment:
I think makes sense.
Normally this code should be triggered by the waiter *because* of a change of state and so should hit the conn->reload() that was already here (at the bottom of the loop).

But if the waiter times out and so loads the state without being triggered, then it could be in a running state etc., but potentially in a race with the session stopping cleanly.